File: app/src/main/java/io/duckduckgosearch/app/AutoCompleteAdapter.java
package io.duckduckgosearch.app;

import android.app.Activity;
import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Arrays;

/**
 * This is a custom adapter for search suggestions list
 */

public class AutoCompleteAdapter extends ArrayAdapter<String> {

    private Context context;
    private int resId;
    private DuckAutoCompleteTextView searchBar;
    private OnItemClickListener clickListener;
    private String[] filteredList;
    private String[] list;
    private ArrayList<String> suggestions;
    private int historyCount;

    public interface OnItemClickListener {
        void onItemClickListener(String searchTerm);
    }

    AutoCompleteAdapter(@NonNull Context context, int resource,
                        @NonNull String[] objects, DuckAutoCompleteTextView searchBar) {
        super(context, resource, objects);
        this.context = context;
        resId = resource;
        list = objects;
        if (list.length >= 5) {
            filteredList = Arrays.copyOf(list, 5, String[].class);
        } else {
            filteredList = Arrays.copyOf(list, list.length, String[].class);
        }
        this.searchBar = searchBar;
        historyCount = filteredList.length;
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                FilterResults results = new FilterResults();

                if (list == null || list.length == 0) {
                    synchronized (this) {
                        results.values = null;
                        results.count = 0;
                    }
                } else {
                    if (charSequence == null || charSequence.length() == 0) {
                        if (list.length < 5) {
                            synchronized (this) {
                                results.values = Arrays.copyOf(list, list.length, String[].class);
                                results.count = list.length;
                            }
                        } else {
                            synchronized (this) {
                                results.values = Arrays.copyOf(list, 5, String[].class);
                                results.count = 5;
                            }
                        }
                    } else {
                        ArrayList<String> matchingTerms = new ArrayList<>();
                        for (String item : list) {
                            if (item.toLowerCase().startsWith(charSequence.toString().toLowerCase())) {
                                matchingTerms.add(item);
                            }
                        }
                        if (matchingTerms.size() < 5) {
                            synchronized (this) {
                                results.values = Arrays.copyOf(matchingTerms.toArray(), matchingTerms.size(), String[].class);
                                results.count = matchingTerms.size();
                            }
                        } else {
                            synchronized (this) {
                                results.values = Arrays.copyOf(matchingTerms.toArray(), 5, String[].class);
                                results.count = 5;
                            }
                        }
                    }
                }

                return results;
            }

            @Override
            protected void publishResults(CharSequence charSequence, final FilterResults filterResults) {
                OnlineACParser parser = new OnlineACParser((String[])filterResults.values);
                parser.setOnParseListener(new OnlineACParser.OnParsed() {
                    @Override
                    public void onParsed(ArrayList<String> list) {
                        suggestions = list;
                        if (filterResults.values == null) {
                            if (suggestions != null && suggestions.size() != 0) {
                                if (suggestions.size() < 5) {
                                    filteredList = Arrays.copyOf(suggestions.toArray(), suggestions.size(), String[].class);
                                } else {
                                    filteredList = Arrays.copyOf(suggestions.toArray(), 5, String[].class);
                                }
                            }
                            filteredList = new String[]{};
                            notifyDataSetInvalidated();
                        } else {
                            if (suggestions != null && suggestions.size() != 0) {
                                ArrayList<String> finalList = new ArrayList<>(Arrays.asList((String[]) filterResults.values));
                                if (filterResults.count < 5) {
                                    for (int i = 0; i < 5 - filterResults.count; i++) {
                                        finalList.add(suggestions.get(i));
                                    }
                                }
                                filteredList = Arrays.copyOf(finalList.toArray(), finalList.size(), String[].class);
                            } else {
                                filteredList = (String[]) filterResults.values;
                            }
                            if (filteredList.length > 0) {
                                notifyDataSetChanged();
                            } else {
                                notifyDataSetInvalidated();
                            }
                        }
                    }
                });
                parser.execute(charSequence.toString());
                historyCount = filterResults.count;
            }
        };
    }

    @Override
    public String getItem(int position) {
        return filteredList[position];
    }

    @Override
    public int getCount() {
        if (filteredList != null) {
            return filteredList.length;
        } else {
            return 0;
        }
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {

        if (convertView == null) {
            LayoutInflater inflater = ((Activity)getContext()).getLayoutInflater();
            convertView = inflater.inflate(resId, parent, false);
        }

        final int pos = position;

        ImageButton appendButton = convertView.findViewById(R.id.auto_complete_item_button);
        appendButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                searchBar.setText("");
                searchBar.append(getItem(pos) + " ");
            }
        });

        RelativeLayout root = convertView.findViewById(R.id.auto_complete_item_root);
        root.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try {
                    clickListener = (OnItemClickListener) context;
                    clickListener.onItemClickListener(getItem(pos));
                    searchBar.setText(getItem(pos));
                } catch (ClassCastException e) {
                    Log.e("AutoCompleteAdapter", "You must implement the clickListenerInterface");
                }
            }
        });
        ImageView icon = convertView.findViewById(R.id.auto_complete_item_icon);
        if (position >= historyCount) {
            icon.setImageDrawable(context.getResources().getDrawable(R.drawable.ic_search_24px));
        } else {
            icon.setImageDrawable(context.getResources().getDrawable(R.drawable.ic_outline_history_24px));
        }
        if (PrefManager.isDarkTheme(context)) {
            appendButton.setImageDrawable(context.getResources().getDrawable(R.drawable.ic_outline_top_left_arrow_white));
            if (position >= historyCount) {
                icon.setImageDrawable(context.getResources().getDrawable(R.drawable.ic_search_24px_white));
            } else {
                icon.setImageDrawable(context.getResources().getDrawable(R.drawable.ic_outline_history_24px_white));
            }
            root.setBackgroundColor(context.getResources().getColor(R.color.darkThemeColorPrimary));
        }

        TextView textItem = convertView.findViewById(R.id.auto_complete_item_text);
        textItem.setText(getItem(position));

        return convertView;
    }
}

File: app/src/main/java/io/duckduckgosearch/app/AutoCompleteAdapter.kt
package io.duckduckgosearch.app

import android.app.Activity
import android.content.Context
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.content.res.ResourcesCompat
import io.duckduckgosearch.app.OnlineACParser.OnParsed
import java.util.*

/**
 * This is a custom adapter for search suggestions list
 */
class AutoCompleteAdapter constructor(context: Context, private val resId: Int,
                                               objects: Array<String>, searchBar: DuckAutoCompleteTextView) : ArrayAdapter<String>(context, resId, objects) {
    private val searchBar: DuckAutoCompleteTextView
    private var clickListener: OnItemClickListener? = null
    private var filteredList: Array<String>?
    private val list: Array<String>?
    private var suggestions: ArrayList<String>? = null
    private var historyCount: Int
    private var parser: OnlineACParser? = null

    interface OnItemClickListener {
        fun onItemClickListener(searchTerm: String?)
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence): FilterResults {
                val results = FilterResults()
                if (list == null || list.isEmpty()) {
                    synchronized(this) {
                        results.values = null
                        results.count = 0
                    }
                } else {
                    if (charSequence.isEmpty()) {
                        if (list.size < 5) {
                            synchronized(this) {
                                results.values = Arrays.copyOf(list, list.size, Array<String>::class.java)
                                results.count = list.size
                            }
                        } else {
                            synchronized(this) {
                                results.values = Arrays.copyOf(list, 5, Array<String>::class.java)
                                results.count = 5
                            }
                        }
                    } else {
                        val matchingTerms: ArrayList<String> = ArrayList()
                        for (item: String in list) {
                            if (item.toLowerCase(Locale.getDefault()).startsWith(charSequence.toString().toLowerCase(Locale.getDefault()))) {
                                matchingTerms.add(item)
                            }
                        }
                        if (matchingTerms.size < 5) {
                            synchronized(this) {
                                results.values = Arrays.copyOf<String, Any>(matchingTerms.toTypedArray(), matchingTerms.size, Array<String>::class.java)
                                results.count = matchingTerms.size
                            }
                        } else {
                            synchronized(this) {
                                results.values = Arrays.copyOf<String, Any>(matchingTerms.toTypedArray(), 5, Array<String>::class.java)
                                results.count = 5
                            }
                        }
                    }
                }
                return results
            }

            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
                if (parser != null && !parser!!.isCancelled) {
                    parser!!.cancel(true)
                }
                parser = OnlineACParser(filterResults.values as Array<String>?)
                parser!!.setOnParseListener(object : OnParsed {
                    override fun onParsed(list: ArrayList<String>?) {
                        suggestions = list
                        if (filterResults.values == null) {
                            if (suggestions != null && suggestions!!.size != 0) {
                                filteredList = if (suggestions!!.size < 5) {
                                    Arrays.copyOf<String, Any>(suggestions!!.toTypedArray(), suggestions!!.size, Array<String>::class.java)
                                } else {
                                    Arrays.copyOf<String, Any>(suggestions!!.toTypedArray(), 5, Array<String>::class.java)
                                }
                                notifyDataSetChanged()
                            } else {
                                filteredList = arrayOf()
                                notifyDataSetInvalidated()
                            }
                        } else {
                            if (suggestions != null && suggestions!!.size != 0) {
                                val finalList: ArrayList<String> = ArrayList(listOf(*filterResults.values as Array<String?>))
                                var count: Int = 5 - filterResults.count
                                if (count > suggestions!!.size) {
                                    count = suggestions!!.size
                                }
                                Log.d("Debug : ", "Suggestions size : " + suggestions!!.size)
                                Log.d("Debug : ", "Difference : $count")
                                if (filterResults.count < 5) {
                                    for (i in 0 until count) {
                                        finalList.add(suggestions!![i])
                                    }
                                }
                                filteredList = Arrays.copyOf<String, Any>(finalList.toTypedArray(), finalList.size, Array<String>::class.java)
                            } else {
                                filteredList = filterResults.values as Array<String>?
                            }
                            if (filteredList!!.isNotEmpty()) {
                                notifyDataSetChanged()
                            } else {
                                notifyDataSetInvalidated()
                            }
                        }
                    }
                })
                parser!!.execute(charSequence.toString())
                historyCount = filterResults.count
            }
        }
    }

    override fun getItem(position: Int): String {
        return filteredList!![position]
    }

    override fun getCount(): Int {
        return if (filteredList != null) {
            filteredList!!.size
        } else {
            0
        }
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var cView: View? = convertView
        if (cView == null) {
            val inflater: LayoutInflater = (context as Activity).layoutInflater
            cView = inflater.inflate(resId, parent, false)
        }
        val pos: Int = position
        val appendButton: ImageButton = cView!!.findViewById(R.id.auto_complete_item_button)
        appendButton.setOnClickListener {
            searchBar.setText("")
            searchBar.append(getItem(pos) + " ")
        }
        val root: RelativeLayout = cView.findViewById(R.id.auto_complete_item_root)
        root.setOnClickListener {
            try {
                clickListener = context as OnItemClickListener?
                clickListener!!.onItemClickListener(getItem(pos))
                searchBar.setText(getItem(pos))
            } catch (e: ClassCastException) {
                Log.e("AutoCompleteAdapter", "You must implement the clickListenerInterface")
            }
        }
        val icon: ImageView = cView.findViewById(R.id.auto_complete_item_icon)
        if (position >= historyCount) {
            icon.setImageDrawable(ResourcesCompat.getDrawable(context.resources,R.drawable.ic_search_24px,null))
        } else {
            icon.setImageDrawable(ResourcesCompat.getDrawable(context.resources,R.drawable.ic_outline_history_24px,null))
        }
        if (PrefManager.isDarkTheme(context)) {
            appendButton.setImageDrawable(ResourcesCompat.getDrawable(context.resources,R.drawable.ic_outline_top_left_arrow_white,null))
            if (position >= historyCount) {
                icon.setImageDrawable(ResourcesCompat.getDrawable(context.resources,R.drawable.ic_search_24px_white,null))
            } else {
                icon.setImageDrawable(ResourcesCompat.getDrawable(context.resources,R.drawable.ic_outline_history_24px_white,null))
            }
            root.setBackgroundColor(ResourcesCompat.getColor(context.resources,R.color.darkThemeColorPrimary,null))
        }
        val textItem: TextView = cView.findViewById(R.id.auto_complete_item_text)
        textItem.text = getItem(position)
        return (cView)
    }

    init {
        list = objects
        filteredList = if (list.size >= 5) {
            Arrays.copyOf(list, 5, Array<String>::class.java)
        } else {
            Arrays.copyOf(list, list.size, Array<String>::class.java)
        }
        this.searchBar = searchBar
        historyCount = (filteredList as Array<String>).size
    }
}